app/src/main/java/org/simple/clinic/pin/PinEntryCardView.kt

```diff
@@ -0,0 +1,117 @@
+package org.simple.clinic.pin
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.support.annotation.StringRes
+import android.support.transition.Fade
+import android.support.transition.TransitionManager
+import android.support.v4.view.animation.FastOutSlowInInterpolator
+import android.support.v7.widget.CardView
+import android.util.AttributeSet
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.Button
+import android.widget.TextView
+import com.jakewharton.rxbinding2.view.RxView
+import io.reactivex.Observable
+import io.reactivex.Single
+import io.reactivex.android.schedulers.AndroidSchedulers.mainThread
+import io.reactivex.schedulers.Schedulers.io
+import io.reactivex.subjects.PublishSubject
+import kotterknife.bindView
+import org.simple.clinic.R
+import org.simple.clinic.activity.TheActivity
+import org.simple.clinic.widgets.StaggeredEditText
+import org.simple.clinic.widgets.hideKeyboard
+import org.simple.clinic.widgets.showKeyboard
+import javax.inject.Inject
+
+class PinEntryCardView(context: Context, attrs: AttributeSet) : CardView(context, attrs) {
+
+  @Inject
+  lateinit var controller: PinEntryCardController
+
+  val pinEditText by bindView<StaggeredEditText>(R.id.pinentry_pin)
+  val forgotPinButton by bindView<Button>(R.id.pinentry_forgotpin)
+
+  private val pinContainer by bindView<ViewGroup>(R.id.pinentry_pin_container)
+  private val progressView by bindView<View>(R.id.pinentry_progress)
+  private val errorTextView by bindView<TextView>(R.id.pinentry_error)
+  private val successfulAuthSubject = PublishSubject.create<PinAuthenticated>()
+
+  sealed class State {
+    object PinEntry : State()
+    object Progress : State()
+  }
+
+  init {
+    LayoutInflater.from(context).inflate(R.layout.pin_entry_card, this, true)
+    moveToState(State.PinEntry)
+  }
+
+  @SuppressLint("CheckResult")
+  override fun onFinishInflate() {
+    super.onFinishInflate()
+    if (isInEditMode) {
+      return
+    }
+    TheActivity.component.inject(this)
+
+    Observable.mergeArray(pinTextChanges())
```

Comment:
We are not handling *IME done* clicks here. If the user submits *IME action done* when they have entered less than the PIN length, the view won't show any feedback.

Reply:
This is intentional. We're not showing any error if you press IME done without entering the full PIN.

Reply:
This will break when you add this widget to the login pin screen then because screens have behaviour on submit clicked.

Reply:
Oh yea, I'll also change that because the new design is to auto-submit PINs everywhere.